picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java

```diff
@@ -0,0 +1,146 @@
+ * program input</li>
+ * <li>Provide a REPL (shell) to process input interactively including
+ * expected editing features at a prompt</li>
+ * </ul>
+ */
+public class ExampleTool {
+    public static void main(String[] args) {
+        exit(new CommandLine(new Options()).execute(args));
+    }
+}
+
+@Command(name = "example tool", mixinStandardHelpOptions = true, version = {"example tool 0-SNAPSHOT"})
+final class Options implements Callable<Integer> {
+    @Option(names = {"--prompt"}, description = {"Change the interactive prompt from '> '."})
+    private String prompt = "> ";
+
+    @Parameters(description = {"Command line arguments", "Edit this help as suitable."})
+    private List<String> arguments = emptyList();
+
+    @Override
+    public Integer call() {
+        // Return failure to demonstrate that `-h` does not fail: standard
+        // options are handled by Picocli: this `call()` method is not
+        // invoked by standard options
+
+        // Use case #1 -- process command line arguments
+        if (!arguments.isEmpty()) {
+            for (final String argument : arguments) {
+                final Integer result = process(argument);
+                if (0 != result) return result;
+            }
+            return 0;
+        }
+        // Use case #2 -- pipe or shell redirection
+        else if (null == System.console()) return runFromInput();
+        // Use case #3 -- interactive REPL
+        else return runRepl(prompt);
+    }
+
+    /** @todo Nicer way using more modern Java. */
+    private static Integer runFromInput() {
+        try (final BufferedReader lineReader = lineReader()) {
+            while (true) {
+                final String line = lineReader.readLine();
+                if (null == line) break;
+                final Integer result = process(line);
+                if (0 != result) return result;
+            }
+        } catch (final IOException e) {
+            return 2;
+        }
+        return 0;
+    }
+
+    private static BufferedReader lineReader() {
+        return new BufferedReader(new InputStreamReader(System.in));
+    }
+}
+
+final class Repl {
+    /**
+     * This is an <em>absolutely minimal</em> REPL.
+     *
+     * @param prompt the prompt to show on left-side asking user for input
+     *
+     * @return the exit code for <code>main()</code>
+     *
+     * @todo Newer versions of Java permit defining this as a lambda
+     */
+    static Integer runRepl(final String prompt) {
+        try (final Terminal terminal = TerminalBuilder.builder().name("example tool").build()) {
+            final LineReader lineReader = LineReaderBuilder.builder().terminal(terminal).build();
+            do {
+                final String input;
+                try {
+                    input = lineReader.readLine(prompt);
+                } catch (final UserInterruptException | EndOfFileException e) {
+                    // Handling user banging on <kbd>Ctrl-C</kbd>
```

Comment:
Do we want to catch these exceptions here?
My preference would be to let this method throw these exceptions and handle them in the place where this method is called.